### PR TITLE
added osqp_eigen_vendor

### PR DESCRIPTION
**Pull request by mhubii**
hey, recently came across this repository and thought it would be helpful to have a vendor for [osqp-eigen](https://github.com/robotology/osqp-eigen), too. This PR is by no means perfect, but maybe something along these lines is desired. The main changes are

- Added umbrella package `osqp_vendors`
- Added `osqp_eigen_vendor`, in accordance to `osqp_vendor`
- Update folder structure accordingly

Thanks

### CODE REVIEW

**Comment by mhubii** [comment]
include directories are not correctly exported #11

**Comment by mhubii** [comment]
header files are not correctly declared #12

**Comment by kenji-miyake** [comment]
@esteve cc @wep21 Could you review this, please? :pray:

File: osqp_vendor/CMakeLists.txt [hunk 1]
```diff
@@ -33,12 +33,12 @@ macro(build_osqp)
     DESTINATION
       ${CMAKE_INSTALL_PREFIX}
     USE_SOURCE_PERMISSIONS
-    PATTERN config.h EXCLUDE
   )
```
Comment (esteve) [inline osqp_vendor/CMakeLists.txt:36]: Why this change?

File: osqp_eigen_vendor/package.xml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>osqp_eigen_vendor</name>
+  <version>0.0.0</version>
+  <description>Vendor package for osqp-eigen https://github.com/robotology/osqp-eigen</description>
```
Comment (esteve) [inline osqp_eigen_vendor/package.xml:6]: package.xml requires an author field, if I'm not mistaken

**Review by esteve — CHANGES_REQUESTED** [review]

File: osqp_vendors/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+cmake_minimum_required(VERSION 3.9)
```
Comment (esteve) [inline osqp_vendors/CMakeLists.txt:1]: There's no need for this package, please remove `osqp_vendors`. Any downstream packages can depend on either `osqp_vendor` or `osqp_eigen_vendor` if needed

File: osqp_vendor/CMakeLists.txt [hunk 4]
```diff
@@ -33,12 +33,12 @@ macro(build_osqp)
...
 endmacro()
 
 build_osqp()
 
+ament_export_include_directories(include/osqp)
```
[the remark this answers is not in the document]
Comment (esteve) [reply osqp_vendor/CMakeLists.txt:41]: Please revert this change, it'll break any downstream package that currently depend on `osqp_vendor`, hierarchy headers are better than flat ones, there's fewer chances of collision.

File: osqp_vendors/package.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  </description>
+  <maintainer email="esteve.fernandez@autoware.org">Esteve Fernandez</maintainer>
+  <license>Apache License 2.0</license>
+
+  <url type="website">https://github.com/oxfordcontrol/osqp</url>
+  <author email="esteve.fernandez@autoware.org">Esteve Fernandez</author>
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
+  <buildtool_depend>git</buildtool_depend>
+
```
[the remark this answers is not in the document]
Comment (esteve) [reply osqp_vendors/package.xml:18]: To answer your questions: I don't think there's any need for `osqp_vendors`, and there should be a separate repository for `osqp_eigen_vendor`, they are separate projects.

File: osqp_vendor/CMakeLists.txt (hunk 1 above)
Comment (esteve) [reply osqp_vendor/CMakeLists.txt:36]: `config.h` is generated at build time, doesn't seem to be the case with `osqp_vendor`, but that's the rationale for excluding it here and why packages don't have it in their source tree.

**Review by esteve — CHANGES_REQUESTED** [review]

File: osqp_vendor/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhubii) [reply osqp_vendor/CMakeLists.txt:41]: I agree, but this is what `osqp` defaults to. Hence not sure how one would resolve this for `osqp_eigen_vendor`

File: osqp_vendors/CMakeLists.txt (hunk 3 above)
Comment (mhubii) [reply osqp_vendors/CMakeLists.txt:1]: removed

File: osqp_vendor/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (esteve) [reply osqp_vendor/CMakeLists.txt:41]: Still, downstream ROS packages expect the `osqp` prefix. For `osqp_eigen_vendor`, I'd patch it so that it includes `osqp` via `osqp/osqp.h`

File: osqp_vendors/package.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (mhubii) [reply osqp_vendors/package.xml]: I'll move `osqp_eigen_vendor` to a separate folder, once the include issue is resolved

File: osqp_vendor/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhubii) [reply osqp_vendor/CMakeLists.txt:41]: does that mean you would patch it in [osqp-eigen](https://github.com/robotology/osqp-eigen). Thanks btw, really appreciate your help

[the remark this answers is not in the document]
Comment (mhubii) [reply osqp_vendor/CMakeLists.txt:41]: I think that PRs in [osqp](https://github.com/osqp/osqp) and [osqp-eigen](https://github.com/robotology/osqp-eigen) are required to fix this. In the meantime, I created PR #13 to fix issue #11.

Please feel free to close this PR. I will take the learnings to create `osqp_eigen_vendor` in a separate repository.

**Comment by mhubii** [comment]
I tried a PR in osqp/osqp#401 to fix this, with little success. The authors of `osqp` do not want to adopt it. One could globally set the include directories in `osqp_eigen_vendor` to `include/osqp`, which is a bit hacky in modern CMake standards.

**Comment by esteve** [comment]
> I tried a PR in [osqp/osqp#401](https://github.com/osqp/osqp/pull/401) to fix this, with little success. The authors of `osqp` do not want to adopt it. One could globally set the include directories in `osqp_eigen_vendor` to `include/osqp`, which is a bit hacky in modern CMake standards.

Thanks for submitting the PR there, it's too bad they don't want to accept the change. Instead, we can patch the vendor package for `osqp-eigen` so that it uses the include path from `osqp_vendor`. However, I just checked and `osqp_vendor` does not seem to have any published dependant packages (https://index.ros.org/p/osqp_vendor/#foxy-deps). We use `osqp_vendor` in Autoware, though.

@wep21 what do you think? Should we export the `osqp.h` header the same way as upstream osqp? (i.e. `#include <osqp.h>` instead of `#include <osqp/osqp.h>`)

**Comment by mhubii** [comment]
in the [develop-1.0](https://github.com/osqp/osqp/tree/develop-1.0) branch, the authors of `osqp` prefix their public headers with `osqp_` instead of exporting `include` so that `osqp/*.h` can be included. Maybe they will reconsider this and it is worth waiting.

File: osqp_vendor/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (traversaro) [reply osqp_vendor/CMakeLists.txt:41]: Hi, speaking as part of the mantainer team of osqp-eigen. : ) 

There is any reason why you do not just call `ament_export_targets`, as document in ament docs: https://docs.ros.org/en/foxy/How-To-Guides/Ament-CMake-Documentation.html#building-a-library ? 

CMake's imported target are explicitly designed to avoid the to manually handle dependencies via `<pkg>_INCLUDE_DIRS` and `<pkg>_LIBRARIES` variables.

[the remark this answers is not in the document]
Comment (mhubii) [reply osqp_vendor/CMakeLists.txt:41]: so the reason I called `ament_export_include_directories` is that `osqp_vendor` doesn't seem to export the includes, for a reason I do not understand. For example, this very simple project won't compile using `osqp_vendor`

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.5)
project(test)

find_package(ament_cmake REQUIRED)
find_package(osqp_vendor REQUIRED)

add_executable(node main.cpp)

# # doesn't work (preferred)
# ament_target_dependencies(node osqp_vendor)

# works
target_link_libraries(node osqp::osqp)

ament_package()

```
main.cpp
```cpp
#include <osqp.h>

int main(int argc, char** argv) {

    return 0;
}
```
If `ament_export_include_directories` is added, then `ament_target_dependencies(node osqp_vendor)` works.

[the remark this answers is not in the document]
Comment (traversaro) [reply osqp_vendor/CMakeLists.txt:41]: Why can't you simply have `find_package(osqp REQUIRED)` instead of `find_package(osqp_vendor REQUIRED)` ? I tought that the whole point of `<pkg>_vendor` packages is just to provide the `<pkg>`  in a colcon/bloom friendly way, nothing more.

[the remark this answers is not in the document]
Comment (mhubii) [reply osqp_vendor/CMakeLists.txt:41]: you can also do `find_package(osqp REQUIRED)` and link via `target_link_libraries(node osqp::osqp)`. But in this case, ament doesn't work at all.

```cmake
cmake_minimum_required(VERSION 3.5)
project(test)

find_package(ament_cmake REQUIRED)
find_package(osqp REQUIRED) 
# works if ament_export_include_directories added
# # find_package(osqp_vendor REQUIRED) (works, if ament_export_include_directories)

add_executable(node main.cpp)

# doesn't work
ament_target_dependencies(node osqp)  

# works if ament_export_include_directories added
# # ament_target_dependencies(node osqp)

ament_package()
```

[the remark this answers is not in the document]
Comment (traversaro) [reply osqp_vendor/CMakeLists.txt:41]: Ah sorry, I got the point now. I think in this part of the code what you should add is `ament_export_targets(osqp::osqp)`, but I am not sure if that would work. Probably you need to first create a new empty interface target, link that one to `osqp::osqp` and then export that one.

[the remark this answers is not in the document]
Comment (traversaro) [reply osqp_vendor/CMakeLists.txt:41]: Something like:
~~~
add_library(osqp_vendor INTERFACE)
target_link_library(osqp_vendor INTERFACE osqp::osqp)
install(
  TARGETS osqp_vendor 
  EXPORT osqp_vendorTargets)
ament_export_targets(osqp_vendorTargets)
~~~

Not sure if this is the preferred way, but probably the best place to discuss this is to open an issue in ament_cmake I guess.

[the remark this answers is not in the document]
Comment (mhubii) [reply osqp_vendor/CMakeLists.txt:41]: thanks for all the feedback, I'll try my best to dig down to the root cause, but I feel my knowledge of what is happening is too limited at this point

**Comment by mhubii** [comment]
> l

this will probably not happen, and it might be `osqp_vendor`'s responsibility to adapt to `osqp`

**Comment by mhubii** [comment]
replaced all `externalproject_add`, as this seems a little easier. But also this PR will probably not go through, hence closing

**Comment by mhubii** [comment]
maybe one more note relating the use of `osqp_vendor`. Looking at this CMake file https://github.com/autowarefoundation/autoware.universe/blob/97590f434e282ee34a174f082385ecd528394a59/common/osqp_interface/CMakeLists.txt#L23, it becomes apparent how difficult it is to use `osqp_vendor` in practise. I think this partially arises from the `osqp` target not being used in accordance with `osqp`

**Comment by mhubii** [comment]
I think due to comments by @traversaro, this might be heading somewhere useful
